core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java

```diff
@@ -153,14 +158,24 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
                 if (instanceMetaDataEntry.getInstanceMetadataType() == null) {
                     if (ig != null) {
                         if (InstanceGroupType.GATEWAY.equals(ig.getInstanceGroupType())) {
-                            if (!primaryIgSelected) {
-                                primaryIgSelected = true;
-                                instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY_PRIMARY);
-                                instanceMetaDataEntry.setServer(Boolean.TRUE);
-                                LOGGER.info("Primary gateway is not selected, let's select this instance: {}", instanceMetaDataEntry.getInstanceId());
+                            if (terminatedPrimaryGwWhichShouldBeRestored.isPresent()) {
+                                if (terminatedPrimaryGwWhichShouldBeRestored.get().getDiscoveryFQDN().equals(instanceMetaDataEntry.getDiscoveryFQDN())) {
```

Comment:
is it a possible scenario we would have an already terminated PGW but don't have IM with the same FQDN or we would have IM with empty FQDN?
If yes this code would leave us without PGW

Reply:
Good catch, I will cover this case.